### PR TITLE
oc cluster up: work around docker attach race condition

### LINKED ISSUE #9470: `oc cluster up` hangs at "checking for available ports"
Following the instructions at https://www.openshift.com/devpreview/register.html I'm running 

```
oc cluster up --version=v1.3.0-alpha.1
```

And this is what I get:

```
-- Checking Docker client ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.3.0-alpha.1 image ... OK
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... ^C
```

until I break it off manually (waited hours at some point, no difference).
##### Version

oc v1.3.0-alpha.1
kubernetes v1.3.0-alpha.1-331-g0522e63
##### Steps To Reproduce
1. run `oc cluster up --version=v1.3.0-alpha.1`
##### Current Result

```
-- Checking Docker client ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.3.0-alpha.1 image ... OK
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... ^C
```
##### Expected Result

More than the current result ;)
##### Additional Information

---

### PR DESCRIPTION
**Pull request by jim-minter**
Fixes #9470

Alternative to #12189

[test]

@csrwng FYI

### CODE REVIEW

File: pkg/bootstrap/docker/run/run.go [hunk 1]
```diff
@@ -192,64 +171,49 @@ func (h *Runner) runWithOutput(stdIn io.Reader, stdOut, stdErr io.Writer) (int,
...
+	}
+	glog.V(5).Infof("Done waiting for container %q, rc=%d", id, rc)
+
+	stdOut, stdErr := &bytes.Buffer{}, &bytes.Buffer{}
+
+	// changed to only reading logs after execution instead of streaming
+	// stdout/stderr to avoid race condition in (at least) docker 1.10-1.14-dev:
+	// https://github.com/docker/docker/issues/29285
+	glog.V(5).Infof("Attaching to container %q", id)
+	err = h.client.AttachToContainer(docker.AttachToContainerOptions{
```
Comment (csrwng) [inline pkg/bootstrap/docker/run/run.go]: Just a question ... why use AttachToContainer and not Logs ?

Comment (jim-minter) [inline pkg/bootstrap/docker/run/run.go]: Only because that's what was there before -- now changed to h.client.Logs(), thanks!

**Comment by csrwng** [comment]
LGTM [merge]

**Comment by bparees** [comment]
@jim-minter I think this is also causing problems in our postcommit hook logic when the postcommit hook operation is very short.  Any chance you can do a PR to fsouza/go-dockerclient to patch the log streaming code there too?  (Assuming i'm correct)
https://github.com/fsouza/go-dockerclient/blob/master/client.go#L510

here's where we're running the container and streaming the logs:
https://github.com/openshift/origin/blob/master/pkg/build/builder/dockerutil.go#L186-L192

**Comment by jim-minter** [comment]
@bparees You might create a new issue for this. Do you have an example/replicator?
At first glance:
- I think it's feasible that https://github.com/docker/docker/issues/29285 could cause a hang at the location you indicate if the hook is short-running.
- I don't think any useful workaround/patch can be sent to fsouza/go-dockerclient.
- I think our options are: a) don't stream container output, only retrieve it after container termination (that's what this PR did for oc cluster up); b) fix (or get fixed) the issue upstream in Docker.

**Comment by bparees** [comment]
@jim-minter opened https://github.com/openshift/origin/issues/12587

the recreate its pretty simple.  define a buildconfig w/ a postcommit hook.  make the postcommit hook be something that's very quick like an echo command.  @spadgett was able to recreate it consistently, i haven't tried myself.

**Pull request merged**